doc/developers/advanced_installation.rst

```diff
@@ -68,7 +68,7 @@ feature, code or documentation improvement).
 
    .. prompt:: bash $
 
-     conda create -n sklearn-env -c conda-forge python numpy scipy cython meson-python ninja
+     conda create -n sklearn-env -c conda-forge python numpy scipy cython meson-python ninja spin
```

Comment:
I'm moving a lot more towards @Micky774 's suggestion of having a way to not copy paste these. Maybe we could at least include an `environment.yml` file?

Reply:
Agreed but not really related to adding spin, do you want to open an issue about this :wink:?

Some possibilities off the top of my head:
- have an environment.yml but then what do you do about pip users? scikit-image seems to have instructions that works for both off the top of my head `conda create -n env python` + `conda install --file` that can use requirements.txt.
- use an environment from the lock-files at least it is tested in the CI on some OS. This may be useful if we ever need to pin some stuff for some time.
- have a spin custom command that installs the build dependencies (personally the one I am always annoyed until I switched to creating environment from lock-files directly is the doc dependencies so many different ones with a mix of conda and pip)

Reply:
Here you could benefit from `pixi` (conda + pypi + lockfile) :)

Reply:
We'll see but for now one baby step at a time :wink:

Reply:
What is making you move away from "copy and paste that command"?

I still like having it and using it. But I'd also like to have `spin setup-me-up` as a shortcut. Mostly because the times when I can't find the above command in my shell history I need to spend 5min digging it out of our docs. So having a shortcut in spin would be useful. Though also a way of fixing a symptom instead of the root cause (why is the command in a place that is hard to find?)

---

No lockfiles for dev envs please. scikit-learn has to work with a large number of version combinations of the (few) dependencies it has. Having different developers on different combinations is IMHO a feature, as it increases the chances of someone noticing that we are making it harder to a combination of versions that isn't "main stream".

Case in point, we recently made it so that you need Numpy 2 to build wheels. Or atleast that is the default. Someone asked about "but do we need to have this?", which at the time I found tedious. But just yesterday I tried to use `numpy<2` to reproduce something and basically failed. But I think it would have been even harder if we didn't have the conversation about "yeah but what about using older versions of Numpy, should that still be allowed/possible?"

Reply:
> What is making you move away from "copy and paste that command"?

I am not sure if this question is for me or for Wolf ... just to be clear you would like to have the full pip command somewhere still mentioned in the documentation? I can certainly do this.

The main argument for `spin` in my mind is that you have discoverability and a simple point of entry for new contributors. So you don't remember the magical command you only have to remember to type `spin` and you'll get some nice info with a few things you can do. Right now this looks like this:
![image](https://github.com/user-attachments/assets/c970fedc-2b7c-4528-a99d-5baf57bceed4)

Also this was probably mentioned before but worth repeating the point, `spin` is quite explicit about the commands it actually runs behind the scenes, this is what you get when you do `spin install -v` so you can still find your command if you want to:
![image](https://github.com/user-attachments/assets/65b77170-99c2-4ed0-bb0c-e4197cfb265e)

My hope right now is to try to keep this issue reasonably focused on adding `spin` to be able to have interested people (like me) experiment with it in their day to day work and see how we like it and what we can make with it.

This is not that easy to keep this PR focused because there are plenty of reasonable approaches that are trying to answer "OK but what is the real problem here, can we not try to fix it instead rather than add yet a new tool".
---

About lock-files for dev environment I would agree with you. I have personally started using the CI conda-lock lock files more recently because I find them convenient to try to reproduce CI issues (and I am on a Linux machine). I would certainly not advocate for "conda-lock files for everyone's dev env" :wink:.

Reply:
Sorry, I was actually thinking of Adrin. Because he said he used to not like the idea but is now changing his mind. So I was interested in what is changing his mind.

I think we should add spin. Like you say it is convenient, for experienced contributors and newcomers.

The point that is important to me is that we treat it as a useful way to have a shortcut and discoverability for very simple steps. Not as a way to add a layer of abstraction that hides/deals with the fact that the actual commands are annoying/hard to understand/there are many of them/etc.

The fact that spin prints what it is doing is good for discoverability, but I think it doesn't provide the forcing function that is needed to keep the things it is doing simple. Building a `Dockerfile` also prints the commands and output, but many of them might as well be written in Latin. I'd understand about the same amount about what they are actually doing.